ogr/ogrsf_frmts/miramon/mm_gdal_functions.c

```diff
@@ -0,0 +1,2801 @@
+    char *pszString;
+
+    if (!szFileName)
+        return 1;
+
+    strcpy(pMMBDXP->szFileName, szFileName);
+    strcpy(pMMBDXP->ReadingMode, "rb");
+
+    if ((pMMBDXP->pfDataBase = fopen_function(pMMBDXP->szFileName,
+                                              pMMBDXP->ReadingMode)) == nullptr)
+        return 1;
+
+    pf = pMMBDXP->pfDataBase;
+
+    fseek_function(pf, 0, SEEK_SET);
+    /* ====== Header reading (32 bytes) =================== */
+    offset_primera_fitxa = 0;
+
+    if (1 != fread_function(&(pMMBDXP->dbf_version), 1, 1, pf) ||
+        1 != fread_function(&variable_byte, 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->month), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->day), 1, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    if (1 != fread_function(&nRecords, 4, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    if (1 != fread_function(&offset_primera_fitxa, 2, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    pMMBDXP->year = (short)(1900 + variable_byte);
+reintenta_lectura_per_si_error_CreaCampBD_XP:
+
+    if (n_queixes_estructura_incorrecta > 0)
+    {
+        if (!MM_ES_DBF_ESTESA(pMMBDXP->dbf_version))
+        {
+            offset_fals = offset_primera_fitxa;
+            if ((offset_primera_fitxa - 1) % 32)
+            {
+                for (offset_fals = (offset_primera_fitxa - 1);
+                     !((offset_fals - 1) % 32); offset_fals--)
+                    ;
+            }
+        }
+    }
+    else
+        offset_reintent = ftell_function(pf);
+
+    if (1 != fread_function(&ushort, 2, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->reserved_1), 2, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->transaction_flag), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->encryption_flag), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->dbf_on_a_LAN), 12, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    if (MM_ES_DBF_ESTESA(pMMBDXP->dbf_version))
+    {
+        memcpy(&pMMBDXP->nRecords, &nRecords, 4);
+        memcpy(((char *)&pMMBDXP->nRecords) + 4, &pMMBDXP->dbf_on_a_LAN, 4);
+    }
+    else
+        pMMBDXP->nRecords = nRecords;
+
+    if (1 != fread_function(&(pMMBDXP->MDX_flag), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->CharSet), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->reserved_2), 2, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    // Checking for a cpg file
+    if (pMMBDXP->CharSet == 0)
+    {
+        FILE_TYPE *f_cpg;
+
+        strcpy(cpg_file, pMMBDXP->szFileName);
+        strcpy(cpg_file, reset_extension(cpg_file, "cpg"));
+        f_cpg = fopen_function(cpg_file, "r");
+        if (f_cpg)
+        {
+            char *p;
+            size_t read_bytes;
+            fseek_function(f_cpg, 0L, SEEK_SET);
+            if (11 > (read_bytes = fread_function(local_message, 1, 10, f_cpg)))
```

Comment:
If I'm not wrong, that test will always success given you ask to read at most 10 bytes.
But what is a bit concerning, and this is a general remark, is the use of the "local_message" variable, which is a static variable. If that can't be eliminated, that means that the Miramon driver is not thread safe. That could cause it to crash if a user for example tries to use it with QGIS, which will generally open a file from multiple threads to speed up rendering.

An option would be to wrap your OGRMiraMonDataSource in a OGRMiraMonDataSource (cf ogr/ogrsf_frmts/generic/ogrmutexeddatasource.h), and use a global mutex for the whole Miramon driver. See ogr/ogrsf_frmts/filegdb/FGdbDriver.cpp for another driver that needs to have a global mutex

Reply:
This local_message can be done as a char[11]. I've done that.

But, about the other local_message... I try to explain.
I wanted to have same messages in GDAL compilation than in MiraMon. I wanted to use functions like CPLError() creating MMCPLError() that in gdal does the same as CPLError() but in MiraMon uses local_message (MiraMon is not thread safe). When I tried to do that I didn't know how to do to manage the lasts arguments in
void CPL_DLL CPLError(CPLErr eErrClass, CPLErrorNum err_no,
                      CPL_FORMAT_STRING(const char *fmt), **...**)

Si, I avoided to use them creating a message before. But I can do it locally and thread safe. 
For me it would be great call MMCPLError() in the same way than CPLError().

I think I am not explaning very well...

Reply:
> For me it would be great call MMCPLError() in the same way than CPLError().

That would mean that for the Miramon specific case, you would need an implementation like
```
void MMCPLError(int code, const char* fmt, ...)
{
    va_list args;
    va_start(args, fmt);
    char szBigEnoughBuffer[1024];
    vsnprintf(szBigEnoughBuffer, sizeof(szBigEnoughBuffer), fmt, args);
    // call here true logging Miramon function with szBigEnoughBuffer
    va_end(args);
}
```
and for the GDAL case, you'd just ``#define MMCPLError CPLError``